### PR TITLE
Homogenizing Dropout

### PR DESCRIPTION
**Pull request by Sinacam**
## What does this PR do?

Added some bert dropout and homogenized naming across all our models.
Dropout can be applied at various points throughout a model, and our previous naming does not reflect this well.
~~This is a work in progress, example configs and tutorials have yet to be changed as well.~~
~~The current default dropout values are arbitrary and should be set to something sensible.~~

Previously, the `encoder_dropout` in RNN/CNN is now called `post_encoder_dropout` because they apply _after_ the RNN/CNN encoder, not within the encoder itself.
RNN now has `encoder_dropout` that applies after each token passes through the model.

Previously, the dropout in BERT models are inconsistent and missing some dropout parameters.
Now we have
+ `encoder_hidden_dropout`
+ `encoder_attention_dropout`
+ `post_encoder_dropout`
+ `attention_dropout`

## Test CLI & API (`bash tests/autotest.sh`)
Test APIs used by main.py.
- [x] Test Pass
  - All tests finished (13/13) on dropout. See test_report.txt for details.
- [ ] Not Applicable (i.e., the PR does not include API changes.)

## Check API Document
If any new APIs are added, please check if the description of the APIs is added to API document. 
- [ ] API document is updated ([linear](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/linear.html), [nn](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/nn.html))
- [x] Not Applicable (i.e., the PR does not include API changes.)

## Test quickstart & API (`bash tests/docs/test_changed_document.sh`)
If any APIs in quickstarts or tutorials are modified, please run this test to check if the current examples can run correctly after the modified APIs are released.

### CODE REVIEW

**Comment by Sinacam** [comment]
Configs and tutorials have been changed.

Bert default dropout values have been changed because it was inconsistent with each other (and looks like a bug).
All configs explicitly set their dropout values so none was affected.

File: libmultilabel/nn/networks/modules.py [hunk 1]
```diff
@@ -102,7 +105,7 @@ class CNNEncoder(nn.Module):
     """
 
     def __init__(
-        self, input_size, filter_sizes, num_filter_per_size, activation, dropout=0, num_pool=0, channel_last=False
+        self, input_size, filter_sizes, num_filter_per_size, activation,  post_encoder_dropout=0, num_pool=0, channel_last=False
```
Comment (Eleven1Liu) [inline libmultilabel/nn/networks/modules.py:108]:
```suggestion
        self, input_size, filter_sizes, num_filter_per_size, activation, post_encoder_dropout=0, num_pool=0, channel_last=False
```

File: libmultilabel/nn/networks/modules.py [hunk 2]
```diff
@@ -75,14 +77,15 @@ class LSTMEncoder(RNNEncoder):
...
 
-    def __init__(self, input_size, hidden_size, num_layers, dropout=0):
-        super(LSTMEncoder, self).__init__(input_size, hidden_size, num_layers, dropout)
+    def __init__(self, input_size, hidden_size, num_layers, encoder_dropout=0, post_encoder_dropout=0):
+        super(LSTMEncoder, self).__init__(input_size, hidden_size, num_layers, encoder_dropout, post_encoder_dropout)
 
-    def _get_rnn(self, input_size, hidden_size, num_layers):
-        return nn.LSTM(input_size, hidden_size, num_layers, batch_first=True, bidirectional=True)
+    def _get_rnn(self, input_size, hidden_size, num_layers, dropout):
+        return nn.LSTM(input_size, hidden_size, num_layers, batch_first=True, dropout=dropout, bidirectional=True)
```
Comment (Eleven1Liu) [inline libmultilabel/nn/networks/modules.py:88]: How about this?
```suggestion
        return nn.LSTM(input_size, hidden_size, num_layers, dropout=dropout, batch_first=True, bidirectional=True)
```

File: libmultilabel/nn/networks/modules.py [hunk 3]
```diff
@@ -58,14 +59,15 @@ class GRUEncoder(RNNEncoder):
...
 
-    def __init__(self, input_size, hidden_size, num_layers, dropout=0):
-        super(GRUEncoder, self).__init__(input_size, hidden_size, num_layers, dropout)
+    def __init__(self, input_size, hidden_size, num_layers, encoder_dropout=0, post_encoder_dropout=0):
+        super(GRUEncoder, self).__init__(input_size, hidden_size, num_layers, encoder_dropout, post_encoder_dropout)
 
-    def _get_rnn(self, input_size, hidden_size, num_layers):
-        return nn.GRU(input_size, hidden_size, num_layers, batch_first=True, bidirectional=True)
+    def _get_rnn(self, input_size, hidden_size, num_layers, dropout):
+        return nn.GRU(input_size, hidden_size, num_layers, batch_first=True, dropout=dropout, bidirectional=True)
```
Comment (Eleven1Liu) [inline libmultilabel/nn/networks/modules.py:70]: How about this?
```suggestion
        return nn.GRU(input_size, hidden_size, num_layers, dropout=dropout, batch_first=True, bidirectional=True)
```

**Review by Eleven1Liu — CHANGES_REQUESTED** [review]

**Comment by Eleven1Liu** [comment]
related to LAAT (RNN)

File: libmultilabel/nn/networks/modules.py (hunk 3 above)
Comment (Sinacam) [reply libmultilabel/nn/networks/modules.py:70]: Is there a reason why one order is preferred?
I followed the ordering in torch docs.

**Review by Eleven1Liu — APPROVED** [review]
